userspace/libscap/scap_udig.c

```diff
@@ -321,6 +332,314 @@ int32_t udig_begin_capture(scap_t* handle, char *error)
+// The following 2 function map the ring buffer and the ring buffer 
+// descriptors into the address space of this process.
+// This is the buffer that will be consumed by scap.
+///////////////////////////////////////////////////////////////////////////////
+int32_t udig_alloc_ring(HANDLE* ring_handle,
+	uint8_t** ring,
+	uint32_t* ringsize,
+	char* error)
+{
+	*ring_handle = NULL;
+
+	//
+	// First, try to open an existing ring
+	//
+	HANDLE fh = OpenFileMapping(
+		FILE_MAP_ALL_ACCESS,
+		FALSE,
+		(TCHAR*)TEXT(UDIG_RING_SM_FNAME));
+
+	if(fh != NULL)
+	{
+		//
+		// Existing ring found, find out the size
+		//
+		uint8_t* pdbuf = (uint8_t*)MapViewOfFile(fh,
+			FILE_MAP_ALL_ACCESS,
+			0,
+			0,
+			0);
+
+		MEMORY_BASIC_INFORMATION info;
+		SIZE_T szBufferSize = VirtualQueryEx(GetCurrentProcess(), pdbuf, &info, sizeof(info));
+
+		*ringsize = (uint32_t)info.RegionSize;
+
+		UnmapViewOfFile(pdbuf);
+	}
+	else
+	{
+		//
+		// No ring found, allocate a new one.
+		// Note that, according to the man page, the content of the buffer will
+		// be initialized to 0.
+		//
+		*ringsize = UDIG_RING_SIZE;
+
+		fh = CreateFileMapping(INVALID_HANDLE_VALUE,
+			NULL,
+			PAGE_READWRITE,
+			0,
+			*ringsize,
+			(TCHAR*)TEXT(UDIG_RING_SM_FNAME));
+
+		if(fh == NULL)
+		{
+			_snprintf(error, SCAP_LASTERR_SIZE, "udig_alloc_ring CreateFileMapping error: %u\n", GetLastError());
+			return SCAP_FAILURE;
+		}
+	}
+
+	//
+	// Map the ring. This is a multi-step process because we want to map two
+	// consecutive copies of the same memory to reuse the driver fillers, which
+	// expect to be able to go past the end of the ring.
+	// First of all, map the first ring copy at exactly the beginning of the 
+	// previously allocated area.
+	*ring = (uint8_t*)MapViewOfFileEx(fh,
+		FILE_MAP_ALL_ACCESS,
+		0,
+		0,
+		*ringsize,
+		NULL);
+	if(*ring == NULL)
+	{
+		snprintf(error, SCAP_LASTERR_SIZE, "can't map first copy of buffer error: %u\n", GetLastError());
+		CloseHandle(fh);
+		return SCAP_FAILURE;
+	}
+
+	// Map the second ring copy just after the end of the first one.
+	uint8_t* buf2 = (*ring) + (*ringsize);
+	uint8_t* ring2 = (uint8_t*)MapViewOfFileEx(fh,
+		FILE_MAP_ALL_ACCESS,
+		0,
+		0,
+		*ringsize,
+		buf2);
+	if(ring2 != buf2)
+	{
+		snprintf(error, SCAP_LASTERR_SIZE, "can't map second copy of buffer, needed %p, obtained %p, base=%p\n", 
+			buf2, ring2, *ring);
+		UnmapViewOfFile(*ring);
+		CloseHandle(fh);
+		return SCAP_FAILURE;
+	}
+
+	*ring_handle = fh;
+	return SCAP_SUCCESS;
+}
+
+int32_t udig_alloc_ring_descriptors(HANDLE* ring_descs_fd, struct ppm_ring_buffer_info** ring_info, 
+	struct udig_ring_buffer_status** ring_status, char *error)
+{
+	*ring_descs_fd = NULL;
+
+	uint32_t mem_size = sizeof(struct ppm_ring_buffer_info) + sizeof(struct udig_ring_buffer_status);
+
+	//
+	// First, try to open an existing memory area
+	//
+	HANDLE fh = OpenFileMapping(FILE_MAP_ALL_ACCESS,
```

Comment:
FTR: This creates a named shared memory object, similarly to how shm open does on linux 

see: https://docs.microsoft.com/en-us/windows/win32/memory/creating-named-shared-memory

Reply:
yup